hrms/payroll/doctype/salary_component/salary_component.py

```diff
@@ -2,14 +2,36 @@
 # For license information, please see license.txt
 
 import frappe
+from frappe import _
 from frappe.model.document import Document
 from frappe.model.naming import append_number_if_name_exists
 
+from hrms.payroll.utils import sanitize_expression
+
 
 class SalaryComponent(Document):
+	def before_validate(self):
+		self._condition, self.condition = self.condition, sanitize_expression(self.condition)
+		self._formula, self.formula = self.formula, sanitize_expression(self.formula)
+
 	def validate(self):
 		self.validate_abbr()
 
+	def on_update(self):
+		# set old values (allowing multiline strings for better readability in the doctype form)
+		if self._condition != self.condition:
+			self.db_set("condition", self._condition)
+		if self._formula != self.formula:
+			self.db_set("formula", self._formula)
+
+	def after_insert(self):
```

Comment:
Move this to validate. People might make changes to the component later like disabling statistical component setting.